src/decoder.rs

```diff
@@ -546,117 +545,143 @@ pub enum ScopeEntry {
     Decoder(Decoder),
 }
 
-pub struct Scope<'a> {
-    parent: Option<&'a Scope<'a>>,
-    names: Vec<Cow<'static, str>>,
-    entries: Vec<ScopeEntry>,
+pub enum Scope<'a> {
+    Empty,
+    Normal(&'a NormalScope<'a>),
```

Comment:
I think it would be better to have `Vec(&'a VecScope<'a>)` or `Multi(&'a MultiScope<'a>)`, as `Normal` doesn't quite explain the trichotomy clearly enough in my view.

Reply:
I've renamed it to MultiScope to parallel SingleScope.